recursion/reading-recursive-code.md

```diff
@@ -0,0 +1,288 @@
+
+When would we choose to solve a problem with recursion? Once we're used to thinking about recursion, recursive algorithms may be more readable and understandable compared to iterative solutions.
+
+## Anatomy of a Recursive Function
+
+Functions that call itself, like the imaginary `get_next_item` function, will recurse an infinite number of times.
+
+However, if a function calls itself an infinite number of times, when will we actually get an answer and solve our problem? Recursive functions are more helpful when they actually _do_ stop!
+
+In order to better read and debug recursive functions, let's dissect the anatomy of one.
+
+### Base Cases are Stopping Conditions
+
+Recursive algorithms will usually take in at least one argument. A recursive algorithm handles different situations based on the value of the argument. The algorithm should handle:
+
+- At least one _recursive case_
+- At least one _base case_
+
+| Part           | Definition                                                            |
+| -------------- | --------------------------------------------------------------------- |
+| Recursive Case | A case or situation that depends on the solution of other cases       |
+| Base Case      | A case or situation where there is a solution without using recursion |
+
+The responsibility of the _recursive case_ is to divide the problem into a smaller problem.
+
+The responsibility of the _base case_ is to give a solid, immediate solution to the problem.
+
+We can imagine that a recursive function will handle recursive cases by calling itself, until it reaches a base case. Most recursive functions will follow this pattern:
+
+1. If some condition is met and we're handling the base case
+   - Solve using the base case's solution
+1. Else
+   - Solve by recursively calling the same function with different input
+
+The "different input" in the recursive call should be working towards bringing the problem to the base case.
+
+## Example: Exponentiation
+
+Read through this recursive function, `pow`, whose responsibility is to calculate [exponentiation](https://en.wikipedia.org/wiki/Exponentiation), otherwise known as `num`<sup>`exponent`</sup>. It takes in two arguments:
+
+1. `num`, which is the base number
+1. `exponent`, which is the exponent
+
+For example, 3<sup>5</sup>, where `3` is `num` and `5` is `exponent`, is expressed and calculated like this:
+
+```
+3 * 3 * 3 * 3 * 3
+```
+
+And 3<sup>6</sup> is expressed as:
+
+```
+3 * 3 * 3 * 3 * 3 * 3
+```
+
+```python
+def pow(num, exponent):
+    if exponent == 0:
+        return 1
+
+    return num * pow(num, exponent - 1)
+```
+
+#### Base Case
+
+In exponentiation, any number raised to the power of zero equals one.
+
+```python
+if exponent == 0:
+    return 1
+```
+
+This code snippet is the _base case_. Whenever `pow` is called **and** the argument `exponent` is `0`, then we return `1`. This base case gives us an immediate solution to the problem: `1`!
```

Comment:
pow to power
I don't like when sentences ending with a number use exclamation points since they look like factorials.
```suggestion
This code snippet is the _base case_. Whenever `power` is called **and** the argument `exponent` is `0`, then we return `1`. This base case gives us an immediate solution to the problem: `1`.
```

Reply:
LOL. this would be the worst case to mistake as a factorial